\documentclass[12pt,a4paper]{article}
\begin{document}
\begin{table}[ht]
\begin{center}
\caption{All statistics are based on contigs of size $\geq$ 500 bp, unless otherwise noted (e.g., "\# contigs ($\geq$ 0 bp)" and "Total length ($\geq$ 0 bp)" include all contigs).}
\begin{tabular}{|l*{2}{|r}|}
\hline
Assembly & meta\_contigs\_1\_to\_meta\_ref\_3\_\_Lactobacillus\_reuteri\_F275\_Kitasato\_55\_65k.fa & meta\_contigs\_2\_to\_meta\_ref\_3\_\_Lactobacillus\_reuteri\_F275\_Kitasato\_55\_65k.fa \\ \hline
\# contigs ($\geq$ 0 bp) & 11 & {\bf 6} \\ \hline
Total length ($\geq$ 0 bp) & {\bf 69110} & 62198 \\ \hline
\# contigs & 11 & {\bf 6} \\ \hline
Largest contig & {\bf 27969} & 27260 \\ \hline
Total length & {\bf 69110} & 62198 \\ \hline
Reference length & 64999 & 64999 \\ \hline
GC (\%) & 40.75 & 40.64 \\ \hline
Reference GC (\%) & 40.73 & 40.73 \\ \hline
N50 & {\bf 17535} & 16099 \\ \hline
NG50 & {\bf 17535} & 16099 \\ \hline
N75 & {\bf 11102} & 10428 \\ \hline
NG75 & {\bf 11102} & 10428 \\ \hline
\# misassemblies & 0 & 0 \\ \hline
Misassembled contigs length & 0 & 0 \\ \hline
\# local misassemblies & 0 & 0 \\ \hline
\# unaligned contigs & 0 + 0 part & 0 + 0 part \\ \hline
Unaligned contigs length & 0 & 0 \\ \hline
Genome fraction (\%) & {\bf 98.397} & 95.437 \\ \hline
Duplication ratio & 1.081 & {\bf 1.003} \\ \hline
\# N's per 100 kbp & 18.81 & {\bf 0.00} \\ \hline
\# mismatches per 100 kbp & 64.11 & {\bf 6.45} \\ \hline
\# indels per 100 kbp & 3.13 & {\bf 0.00} \\ \hline
Largest alignment & {\bf 27969} & 27260 \\ \hline
NA50 & {\bf 17534} & 16099 \\ \hline
NGA50 & {\bf 17534} & 16099 \\ \hline
NA75 & {\bf 11101} & 10428 \\ \hline
NGA75 & {\bf 11101} & 10428 \\ \hline
\end{tabular}
\end{center}
\end{table}
\end{document}
